src/main/sd/init.c

```diff
@@ -0,0 +1,345 @@
+
+    uint8_t sd_next_command;
+    uint8_t buffer[] = {0xff, 0xff, 0xff, 0xff, 0xff, 0xff};
+
+    while(1)
+    {
+        switch(sd_state) 
+        {
+            case SD_IDLE:
+            {
+                if(sd_card_initialized) sd_state = SD_INITIALIZED;
+                    else sd_state = SD_START_INIT;
+                break;
+            }
+
+            case SD_START_INIT:
+            {
+                if(mext2_configure_pins() == 1)
+                {
+                    debug("Error: pin initialization failed\n");
+                    sd_state = SD_ERROR;
+                }
+                else
+                {
+                    usleep(1000);  //wait 1ms
+                    digitalWrite(MEXT2_MOSI, MEXT2_HIGH);
+
+                    for(uint8_t i = 0; i < 74; i++) //74 dummy clock cycles
+                    {
+                        digitalWrite(MEXT2_SCLK, MEXT2_HIGH);
+                        digitalWrite(MEXT2_SCLK, MEXT2_LOW);
+                    }
+
+                    command.index = calc_command_number(COMMAND_SOFTWARE_RESET);
+                    command.argument[] = {0x00, 0x00, 0x00, 0x00};
+                    command.crc = calc_crc7(command, COMMAND_SIZE - sizeof(uint8_t));
+
+                    sd_state = SD_SEND_COMMAND;
+                    sd_next_command = SD_PROCESS_CMD0;
+                }
+                break;
+            }
+
+            case SD_SEND_COMMAND:
+            {
+                mext2_spi_read_write(command, COMMAND_SIZE);
+
+                buffer[] = {0xff, 0xff, 0xff, 0xff, 0xff, 0xff};
+
+                uint16_t i;
+                for(i; i < N_CYCLES_TIMEOUT && buffer[0] == 0xff; i++)
+                {
+                    buffer[0] = 0xff;
+                    mext2_spi_read_write(buffer, 1)
+                }
+
+                if(i == N_CYCLES_TIMEOUT)
+                {
+                    debug("Error: exceeded time limit waiting for response, check your SD card reader device.\n");
+                    mext2_reset_pins();
+                    sd_state = SD_ERROR;
+                }
+                else
+                {
+                    sd_state = sd_next_command;
+                }
+
+                break;
+            }
+
+            case SD_PROCESS_CMD0:
+            {
+                if(buffer[0] != R1_IDLE_STATE)
+                {
+                    debug("Error: not an SD card\n");
+                    mext2_reset_pins();
+                    sd_state = SD_ERROR;
+                }
+                else
+                {
+                    buffer[0] = 0xff;
+                    mext2_spi_read_write(buffer, 1);
+                    
+                    //set CMD8
+                    command.index = calc_command_number(COMMAND_CHECK_VOLTAGE);
+                    command.argument[] = {0x00, 0x00, 0x01, 0xaa};
+                    command.crc = calc_crc7(command, COMMAND_SIZE - sizeof(uint8_t));
```

Comment:
This code is often repeated - maybe we should add a special function for sending command and getting response as return value? (It could help us reduce this switch case SD_SEND_COMMAND - it would be in function in sd_common)